### ISSUE TITLE
fatal error: 40: no cipher suites in common enabling SSL

### ISSUE THREAD

**Issue by RobertoBoni67** [reactor/reactor-netty#1576]
I have to enable SSL in a spring reactive application using Netty as HTTPS server.

For test purpose, I made a very simple spring boot reactive application with a very simple rest controller that simply returns "Hello World".

POM:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 https://maven.apache.org/xsd/maven-4.0.0.xsd">
	<modelVersion>4.0.0</modelVersion>
	<parent>
		<groupId>org.springframework.boot</groupId>
		<artifactId>spring-boot-starter-parent</artifactId>
		<!--version>2.4.4</version-->
		<version>2.2.1.RELEASE</version>
		<relativePath/> <!-- lookup parent from repository -->
	</parent>
	<groupId>hello</groupId>
	<artifactId>reactive-rest-service</artifactId>
	<version>0.0.1-SNAPSHOT</version>
	<name>reactive-rest-service</name>
	<description>Reacror Netty Hello</description>
	<properties>
		<java.version>1.8</java.version>
	</properties>
	<dependencies>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-webflux</artifactId>
		</dependency>
		<dependency>
			<groupId>org.springframework.boot</groupId>
			<artifactId>spring-boot-starter-test</artifactId>
			<scope>test</scope>
		</dependency>
		<dependency>
			<groupId>io.projectreactor</groupId>
			<artifactId>reactor-test</artifactId>
			<scope>test</scope>
		</dependency>
	</dependencies>
	<build>
		<plugins>
			<plugin>
				<groupId>org.springframework.boot</groupId>
				<artifactId>spring-boot-maven-plugin</artifactId>
			</plugin>
		</plugins>
	</build>
</project>
```

Application properties:
```
spring.application.name=ReactorNettyHello

server.port=8282
server.ssl.enabled: true
server.ssl.client-auth: none
server.ssl.key-password: myPassword
server.ssl.key-alias: MyKeystore
server.ssl.key-store: classpath:keystore/my_server_keystore.jks
server.ssl.key-store-password: myPassword
server.ssl.key-store-provider: SUN
server.ssl.key-store-type: JKS
server.ssl.trust-store: classpath:keystore/my_server_truststore.jks
server.ssl.trust-store-password: myPassword
server.ssl.trust-store-provider: SUN
server.ssl.trust-store-type: JKS

logging.level.reactor.netty=INFO
logging.level.org.springframework.http.server.reactive=DEBUG
```

I get the error "fatal error: 40: no cipher suites in common" in SSL handshake while connecting to the Netty server even from a browser (chrome and firefox).

Here is the full trace of the SSL handshake:
```
Using SSLEngineImpl.
reactor-http-nio-3, called closeOutbound()
reactor-http-nio-3, closeOutboundInternal()
Allow unsafe renegotiation: false
Allow legacy hello messages: true
Is initial handshake: true
Is secure renegotiation: false
reactor-http-nio-2, called closeInbound()
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 for TLSv1
reactor-http-nio-2, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
Using SSLEngineImpl.
reactor-http-nio-3, called closeOutbound()
Allow unsafe renegotiation: false
Allow legacy hello messages: true
Is initial handshake: true
Is secure renegotiation: false
reactor-http-nio-2, called closeOutbound()
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 for TLSv1
reactor-http-nio-2, closeOutboundInternal()
reactor-http-nio-3, closeOutboundInternal()
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 for TLSv1.1
reactor-http-nio-2, called closeOutbound()
reactor-http-nio-3, called closeInbound()
reactor-http-nio-2, closeOutboundInternal()
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 for TLSv1.1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 for TLSv1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 for TLSv1.1
reactor-http-nio-3, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384 for TLSv1.1
Ignoring unsupported cipher suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 for TLSv1.1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 for TLSv1.1
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 for TLSv1.1
reactor-http-nio-3, called closeOutbound()
reactor-http-nio-5, READ: TLSv1 Handshake, length = 512
Ignoring unsupported cipher suite: TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384 for TLSv1.1
reactor-http-nio-3, closeOutboundInternal()
*** ClientHello, TLSv1.2
reactor-http-nio-3, called closeOutbound()
reactor-http-nio-3, closeOutboundInternal()
reactor-http-nio-4, READ: TLSv1 Handshake, length = 512
RandomCookie:  GMT: -1597875481 bytes = { 186, 40, 95, 126, 112, 92, 37, 203, 165, 105, 87, 38, 44, 101, 202, 224, 16, 131, 24, 213, 241, 215, 243, 81, 221, 82, 186, 174 }
Session ID:  *** ClientHello, TLSv1.2
{70, 220, 237, 31, 76, 114, 178, 103, 178, 223, 17, 23, 42, 180, 97, 233, 209, 109, 6, 146, 167, 159, 218, 247, 236, 66, 147, 152, 115, 24, 195, 126}
RandomCookie:  Cipher Suites: [Unknown 0x7a:0x7a, Unknown 0x13:0x1, Unknown 0x13:0x2, Unknown 0x13:0x3, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA2
56, TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, Unknown 0xcc:0xa9, Unknown 0xcc:0xa8, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES
_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_GCM_SHA384, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, SSL_RSA_WITH_3DES_EDE_CBC_SHA]
GMT: 1016038084 bytes = { Compression Methods:  { 97, 204, 47, 30, 76, 96, 6, 210, 244, 10, 229, 13, 182, 111, 2, 192, 6, 156, 176, 251, 201, 105, 135, 171, 231, 67, 243, 191 }
0Session ID:   }
{69, 37, 165, 161, 2, 68, 242, 239, 13, 160, 70, 47, 56, 119, 82, 221, 63, 137, 61, 105, 48, 5, 154, 248, 164, 150, 0, 204, 54, 221, 67, 145}
Unsupported extension type_6682, data:
Cipher Suites: [Unknown 0xca:0xca, Unknown 0x13:0x1, Unknown 0x13:0x2, Unknown 0x13:0x3, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_E
CDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, Unknown 0xcc:0xa9, Unknown 0xcc:0xa8, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, T
LS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_256_GCM_SHA384, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA, SSL_RSA_WITH_3DES_EDE_CBC_SHA]
Compression Methods:  { Extension server_name, server_name: [type=host_name (0), value=localhost]
0 }
Extension extended_master_secret
Unsupported extension type_47802, data:
Extension renegotiation_info, renegotiated_connection: <empty>
Extension server_name, server_name: [type=host_name (0), value=localhost]
Extension extended_master_secret
Extension elliptic_curves, curve names: {unknown curve 6682, unknown curve 29, secp256r1, secp384r1}
Extension renegotiation_info, renegotiated_connection: <empty>
Extension ec_point_formats, formats: [uncompressed]
Extension elliptic_curves, curve names: {unknown curve 47802, unknown curve 29, secp256r1, secp384r1}
Unsupported extension type_35, data:
Extension ec_point_formats, formats: [uncompressed]
Unsupported extension type_16, data: 00:0c:02:68:32:08:68:74:74:70:2f:31:2e:31
Unsupported extension type_35, data:
Unsupported extension status_request, data: 01:00:00:00:00
Unsupported extension type_16, data: 00:0c:02:68:32:08:68:74:74:70:2f:31:2e:31
Unsupported extension status_request, data: 01:00:00:00:00
Extension signature_algorithms, signature_algorithms: SHA256withECDSA, Unknown (hash:0x8, signature:0x4), SHA256withRSA, SHA384withECDSA, Unknown (hash:0x8, signature:0x5), SHA384w
ithRSA, Unknown (hash:0x8, signature:0x6), SHA512withRSA, SHA1withRSA
Extension signature_algorithms, signature_algorithms: SHA256withECDSA, Unknown (hash:0x8, signature:0x4), SHA256withRSA, SHA384withECDSA, Unknown (hash:0x8, signature:0x5), SHA384w
ithRSA, Unknown (hash:0x8, signature:0x6), SHA512withRSA, SHA1withRSA
Unsupported extension type_18, data:
Unsupported extension type_18, data:
Unsupported extension type_51, data: 00:29:ba:ba:00:01:00:00:1d:00:20:a6:c0:bb:b9:62:b7:82:d4:78:a8:e8:e0:12:13:8f:60:39:64:49:ab:66:9a:0c:28:dd:ca:d0:9f:3d:7e:7c:3e
Unsupported extension type_51, data: 00:29:1a:1a:00:01:00:00:1d:00:20:f0:5e:7d:96:57:82:21:c5:c1:2e:14:e3:20:ad:cb:95:a4:78:36:30:c2:5c:c6:96:b0:dd:ff:15:42:04:a6:11
Unsupported extension type_45, data: 01:01
Unsupported extension type_45, data: 01:01
Unsupported extension type_43, data: 0a:5a:5a:03:04:03:03:03:02:03:01
Unsupported extension type_43, data: 0a:6a:6a:03:04:03:03:03:02:03:01
Unsupported extension type_27, data: 02:00:02
Unsupported extension type_43690, data: 00
Unsupported extension type_21, data: 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
Unsupported extension type_27, data: 02:00:02
Unsupported extension type_10794, data: 00
Unsupported extension type_21, data: 00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00:00
%% Initialized:  [Session-11, SSL_NULL_WITH_NULL_NULL]
reactor-http-nio-4, fatal error: 40: no cipher suites in common
javax.net.ssl.SSLHandshakeException: no cipher suites in common
%% Initialized:  [Session-12, SSL_NULL_WITH_NULL_NULL]
%% Invalidated:  [Session-11, SSL_NULL_WITH_NULL_NULL]
reactor-http-nio-4, SEND TLSv1.2 ALERT:  fatal, description = handshake_failure
reactor-http-nio-5, fatal error: 40: no cipher suites in common
javax.net.ssl.SSLHandshakeException: no cipher suites in common
reactor-http-nio-4, WRITE: TLSv1.2 Alert, length = 2
%% Invalidated:  [Session-12, SSL_NULL_WITH_NULL_NULL]
reactor-http-nio-4, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLHandshakeException: no cipher suites in common
reactor-http-nio-5, SEND TLSv1.2 ALERT:  fatal, description = handshake_failure
reactor-http-nio-4, called closeOutbound()
reactor-http-nio-5, WRITE: TLSv1.2 Alert, length = 2
reactor-http-nio-4, closeOutboundInternal()
reactor-http-nio-5, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLHandshakeException: no cipher suites in common
reactor-http-nio-5, called closeOutbound()
reactor-http-nio-5, closeOutboundInternal()
reactor-http-nio-4, called closeOutbound()
reactor-http-nio-4, closeOutboundInternal()
reactor-http-nio-4, called closeInbound()
reactor-http-nio-4, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
reactor-http-nio-5, called closeOutbound()
reactor-http-nio-5, closeOutboundInternal()
reactor-http-nio-5, called closeInbound()
reactor-http-nio-4, called closeOutbound()
reactor-http-nio-5, fatal: engine already closed.  Rethrowing javax.net.ssl.SSLException: Inbound closed before receiving peer's close_notify: possible truncation attack?
reactor-http-nio-4, closeOutboundInternal()
reactor-http-nio-5, called closeOutbound()
reactor-http-nio-4, called closeOutbound()
reactor-http-nio-5, closeOutboundInternal()
reactor-http-nio-4, closeOutboundInternal()
reactor-http-nio-5, called closeOutbound()
reactor-http-nio-5, closeOutboundInternal()
```

I use self signed certificates made with openssl and then converted in JKS.

I tryed different versions of spring-boot and the latest that works is spring-boot 2.2.1.RELEASE: with this versions the SSL handshake is successfully completed and all is working fine, but with any newer versions I tryed, I always get the no cipher suites in common error.

I also used nmap to do an SSL scan and it reports certificates and ciphers correctly with the working versions, but it doesn't report anything with the versions that gives the no ciphers error... it seems that that this server dosn't have any chipher available...

It seems a problem related to Netty and SSL...

I am using "jdk1.8.0_161", but I also used java 11 with the same result...

Does anyone have a hint for this behaviour?

**Comment by violetagg** [comment]
@RobertoBoni67 One the server if you enable the logging `logging.level.reactor.netty.debug`, is there more information in the stack trace?

**Comment by RobertoBoni67** [comment]
Hi Violeta.

I did my tests again with the following POM:

**Comment by violetagg** [comment]
@RobertoBoni67 Can you also enable `logging.level.io.netty` and check for these logs:

```
io.netty.handler.ssl.JdkSslContext       : Default protocols (JDK):
io.netty.handler.ssl.JdkSslContext       : Default cipher suites (JDK):
```

**Comment by RobertoBoni67** [comment]
Hi Violeta, and thanks for your support.

I added "logging.level.io.netty=DEBUG" in my app properties.

thr whole log for spring boot 2.2.1 and 2.4.4 are attached:
[log-2.2.1.txt](https://github.com/reactor/reactor-netty/files/6263644/log-2.2.1.txt)
[log-2.4.4.txt](https://github.com/reactor/reactor-netty/files/6263645/log-2.4.4.txt)

Anyway I see these line in both log files:

io.netty.handler.ssl.OpenSsl             : netty-tcnative not in the classpath; OpenSslEngine will be unavailable.
io.netty.handler.ssl.JdkSslContext       : Default protocols (JDK): [TLSv1.2, TLSv1.1, TLSv1] 
io.netty.handler.ssl.JdkSslContext       : Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA]

I can't give a meaning to the first line (the one about OpenSsl engine), but the protocols and cipher suites are the same for both spring versions.

Anyway with spring boot 2.2.1 the SSL handshake is done, while with spring boot 2.4.4 not...

**Comment by violetagg** [comment]
@RobertoBoni67 Can you try Spring Boot `2.4.5-SNAPSHOT`? I would like to be sure that you do not face these two bugs
#1543
https://github.com/spring-projects/spring-boot/issues/25913

**Comment by RobertoBoni67** [comment]
Hi Violeta.

I tryed Spring Boot 2.4.5-SNAPSHOT with the same result...

Here it the whole log
[log-2.4.5-SNAPSHOT.txt](https://github.com/reactor/reactor-netty/files/6283965/log-2.4.5-SNAPSHOT.txt)

**Comment by RobertoBoni67** [comment]
Hi Violeta.

I have some updates.
I did some tests on this issue.

1) The first thing I did was to change Netty with Tomcat in the POM.xml file. With this change I got the SSL handshake working!

2) Looking at this atricle "https://www.baeldung.com/spring-boot-reactor-netty", I found tha I can configure SSL in Nett with a class like the following:

**Comment by violetagg** [comment]
@RobertoBoni67 I'm not familiar with this code in Spring Boot, but it seems that there isn't issue in Reactor Netty, is that correct?

**Comment by RobertoBoni67** [comment]
Hi Violeta.

Finally I found the source of the issue.

It is definitely NOT an issue o r a bug in reactor netty!

But my problem was only done to the way spring boot handle ssl properties fron the application.properties file...

I am sorry, but I was relaying on what worked for Tomcat and also I found that till spring boot 2.2.1 the same SSL configuration is working for both Tomcat and Netty.

I found the source of the issue looking at the source code of the class "SslServerCustomizer.java", that is a class used internally by spring boot to set SSL configuration in Netty.

I found that setting the property "server.ssl.key-alias" makes the SslServerCustomizer class to get a KeyManagerFactory and so a KeyManager in a way that I guess fails to set cipher alghorithms... but I don't understand well why... maybe it's needed some further configuration that I don't know...

For Tomcat the property server.ssl.key-alias seems to be ignored.

And till spring boot 2.2.1 the property server.ssl.key-alias is ignored by the SslServerCustomizer class and this is the reason why all is working till spring boot 2.2.1. From 2.2.1 the class SslServerCustomizer was changed to make other things if the property server.ssl.key-alias is not null...

In conclusion the solution of my issue was indeed very simple: cut away the propery server.ssl.key-alias (so it is null in the class SslServerCustomizer) and all is working great!!!

Sorry for my misunderstanding... maybe I didn't find or dindn't notice something in the documentation...

So you can close this issue.

Thanks very much for your support.

Best Regards.

Roberto.

**Comment by violetagg** [comment]
@RobertoBoni67 Thanks for clarifying the root cause!

**Comment by jonathan-vandebosch** [comment]
So , I also invetigated the issue...
The issue is that at a moment with the alias the private key will be retrieved... and guess what ?
IN LOWERCASES... so if you put your alias with some uppercases in it it will never work !

 
` 
// class SunX509KeyManagerImpl
public PrivateKey getPrivateKey(String var1) {
        if (var1 == null) {
            return null;
        } else {
            X509Credentials var2 = (X509Credentials)this.credentialsMap.get(var1);  <---- lower cases
            return var2 == null ? null : var2.privateKey;
        }
    }
   // class PKCS12KeyStore 
    public boolean engineContainsAlias(String var1) {
        return this.entries.containsKey(var1.toLowerCase(Locale.ENGLISH));
    }

`
